pytorch_lightning/core/lightning.py

```diff
@@ -1724,26 +1736,37 @@ def module_arguments(self) -> dict:
 
 
 def _collect_init_args(frame, path_args: list) -> list:
-    """Recursive search for all children."""
-    if '__class__' in frame.f_locals:
-        local_args = dict(frame.f_locals)
-        local_args.update(local_args.get('kwargs', {}))
-        local_args = {k: v for k, v in local_args.items()
-                      if k not in ('args', 'kwargs', 'self', '__class__', 'frame', 'frame_args')}
-        # if 'hparams' in local_args:
-        #     # back compatible hparams as single argument
-        #     hparams = local_args.get('hparams')
-        #     local_args.update(vars(hparams) if isinstance(hparams, Namespace) else hparams)
+    """
+    Recursively collects the arguments passed to the child constructors in the inheritance tree.
+
+    Args:
+        frame: the current stack frame
+        path_args: a list of dictionaries containing the constructor args in all parent classes
+
+    Return:
+          A list of dictionaries where each dictionary contains the arguments passed to the
+          constructor at that level. The last entry corresponds to the constructor call of the
+          most specific class in the hierarchy.
+    """
+    _, _, _, local_vars = inspect.getargvalues(frame)
+    if '__class__' in local_vars:
+        cls = local_vars['__class__']
+        spec = inspect.getfullargspec(cls.__init__)
+        init_parameters = inspect.signature(cls.__init__).parameters
+        self_identifier = spec.args[0]        # "self" unless user renames it (always first arg)
+        varargs_identifier = spec.varargs     # by convention this is named "*args"
+        kwargs_identifier = spec.varkw        # by convention this is named "**kwargs"
+        exclude_argnames = (
+            varargs_identifier, kwargs_identifier, self_identifier, '__class__', 'frame', 'frame_args'
```

Comment:
why also ignore `*args` and `**kwargs`?

Shouldn't we collect them as well and use them for reinstantiation/loading? Otherwise the module might be instantiated with different params which could cause incompatibilities with module and loaded checkpoint.

Reply:
Because if we have this class
```python
class B(object):

    def __init__(self, *args, **kwargs):
        super().__init__()
        something = kwargs.get('something')
        self.auto_collect_arguments()
```
and we call it like this
`B(something=1, other=2)`
then we want to module arguments to be 
`dict(something=1, other=2)` 
and not
`dict(args=[], kwargs={something=1, other=2})`

It was like this before, this has not changed. The code that I added there just makes it so that the name *args and **kwargs is not hardcoded, but can be named whatever the user wants. 
The Pyhton inspection magic makes it possible to determine which is which.

Reply:
That's the explanation why the ignore, and the second part of your question:
The args inside **kwargs are still collected, see this line
`local_args.update(local_args.get('kwargs', {}))`